### ISSUE TITLE
Runner Hangs When I Stream Two Different Reports To A File

### ISSUE THREAD

**Issue by bhreinb** [DevExpress/testcafe#3201]
### What is your Test Scenario?
I run TestCafe via the programming interface (https://devexpress.github.io/testcafe/documentation/using-testcafe/programming-interface/)
however, when I stream two or more reports to a file I find that the test runner hangs.
 
### What is the Current behavior?
When two or more reports are streamed to a file the test runner hangs.
 
### What is the Expected behavior?
When two or more reports are streamed to file they should be wrote to the filesystem as per normal.
 
### What is your web application and your TestCafe test code?
```
  static main() {
    const testCafeRunner = new TestCafeRunner();
    return testCafe()
      .then((tc) => {
        testCafeServer = tc;
        const tcRunner = testCafeServer.createRunner();
        TestCafeRunner.addScreenShots(tcRunner, 'screenshots');
        return tcRunner
          .src(testCafeRunner.fetchTheSuiteMap())
          .browsers(testCafeRunner.parseTheBrowserConfig())
          .reporter('spec')
          .reporter(reporterCucumberJson, createWriteStream('/AUTOBAHN/reports/reporterCucumberJson/test.json'))
          .reporter(reporterHtmlTestRail, createWriteStream('/AUTOBAHN/reports/reporterHtmlTestRail/test.html'))
          .run({
            skipJsErrors: true,
            skipUncaughtErrors: true,
            pageLoadTimeout: 10000,
          });
      })
      .then((failedCount) => {
        console.log(`Tests failed: ${failedCount}`);
        return testCafeServer.close();
      })
      .catch((error) => {
        console.log(`An Error occurred ${error}`);
      });
  }
```

### Steps to Reproduce:
So I'm running TestCafe with the aid of the following links:
https://testcafe-discuss.devexpress.com/t/run-testsuite-on-remote/423
https://devexpress.github.io/testcafe/documentation/using-testcafe/programming-interface/
 
when I run the program with reporter spec and another reporter the program completes as per normal. However, when I have the spec reporter with two reporters **that stream to a file** I find that the runner never gets to to this point

```
      .then((failedCount) => {
        console.log(`Tests failed: ${failedCount}`);
        return testCafeServer.close();
      })
```

namely the runner hangs.
 
### Your Environment details:
testcafe version: 0.23.2
node.js version: 8.11.3
browser name and version: Chrome 70.
platform and version: Mac OS X 10.13.6

I also found this issue https://github.com/DevExpress/testcafe/issues/2753 (I wonder are they related anyway).

**Comment by miherlosev** [comment]
Hi @bhreinb

Unfortunately, I cannot reproduce the problem with your steps.
I've created an [example](https://github.com/DevExpress/testcafe/files/2663335/i3201.zip) to demonstrate my attempt.
I guess the cause of the problem is in the use of non built-in reporters (`reporterCucumberJson`, `reporterHtmlTestRail`).

Please modify your sample or my project so that I can reproduce the problem on my side.

**Comment by chheller** [comment]
I can verify this exists in version 0.23.2 and in 0.23.3-alpha.2 as I'm currently using the alpha build to get around issue #2916 which seems to be occurring fairly frequently in versions prior to the alpha for our tests. However, the alpha build version 0.23.2 both have this bug where specifying multiple reporters using custom streams causes the process to hang. Using `stdout` and `null` for the streams with multiple reporters does not cause this bug, though all reporters will write to stdout which is an unrelated issue. 

I've created a repo here to demonstrate this bug similarly to how we've implemented Testcafe. It's a minimalistic reproduction and it uses the testcafe API to create a runner, and custom Transform streams to pipe data from the reporter to `stdout` when all the tests are complete
https://github.com/chheller/testcafe-stream-bug

**Comment by miherlosev** [comment]
Hi @chheller
Thank you for the provided example.
At first sight, we cannot determine the source of the problem and need some time to thoroughly investigate your example.
We will get back to you once we make any progress.

**Comment by AndreyBelym** [comment]
@chheller the cause of the problem with the `OutStream` class is that it calls callbacks inside stream methods synchronously. In version `0.23.2` and above, TestCafe tries to properly close reporter streams but assigns necessary event handlers after calling the `stream.end()` method.   I think that it's better to qualify it as a separate bug #3209, because @bhreinb uses the `fs.createWriteStream` function that creates streams that call their callbacks asynchronously.

As a workaround, you can invoke callbacks in your `OutStream` class in an asynchronous way using the `setTimeout` function:
```
class OutStream extends Transform {
  _transform(chunk, encoding, next) {
    console.log(chunk);
    this.push(chunk, encoding);
    setTimeout(next, 0);
  }
}
```

**Comment by JamieMason** [comment]
Apologies if this is not relevant but I _think_ my issue is the same as this one and, from what I can tell, it is an issue with the JSON reporter.

I think this because when running `testcafe --concurrency 1 -r spec 'chrome:headless --allow-insecure-localhost' ./some.ui.spec.js` testcafe will run the one test and exit as expected, but when running `testcafe --concurrency 1 -r spec,json:/path/to/test-report.json 'chrome:headless --allow-insecure-localhost' ./some.ui.spec.js`, testcafe will run the one test – but not exit. 

The output is the same, a successful test run, but the Terminal is not released back to the user.

I am using testcafe 0.23.2 in Node 8.11.1 on macOS 10.12.6.

**Comment by JamieMason** [comment]
Digging into the commit history for testcafe@0.23.2 I found this recent commit 6f8db5e which relates to waiting for streams, so I tried downgrading to testcafe@0.23.1 and can confirm that fixes _my_ issue. I don't know but I think there may have been a regression in 6f8db5e unfortunately @AndreyBelym.

It might be worth also trying your issue with testcafe@0.23.1 @bhreinb.

**Comment by AndreyBelym** [comment]
@JamieMason, thank you, I ran TestCafe several times on macOS with the `-r json:report.txt` argument and noticed that sometimes node.js can fire the `finish` event before our code starts listening to it in the Promise constructor. 

It means that issue #3209 actually describes the problem so I close this issue as a duplicate but will try to fix #3209 right now and post a notification in this thread when a fix is available.
